x-pack/plugin/esql/src/test/java/org/elasticsearch/xpack/esql/expression/function/scalar/convert/ToDatePeriodTests.java

```diff
@@ -0,0 +1,92 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.esql.expression.function.scalar.convert;
+
+import com.carrotsearch.randomizedtesting.annotations.Name;
+import com.carrotsearch.randomizedtesting.annotations.ParametersFactory;
+
+import org.apache.lucene.util.BytesRef;
+import org.elasticsearch.xpack.esql.core.expression.Expression;
+import org.elasticsearch.xpack.esql.core.tree.Source;
+import org.elasticsearch.xpack.esql.core.type.DataType;
+import org.elasticsearch.xpack.esql.expression.function.AbstractFunctionTestCase;
+import org.elasticsearch.xpack.esql.expression.function.FunctionName;
+import org.elasticsearch.xpack.esql.expression.function.TestCaseSupplier;
+import org.elasticsearch.xpack.esql.type.EsqlDataTypeConverter;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Supplier;
+
+@FunctionName("to_dateperiod")
+public class ToDatePeriodTests extends AbstractFunctionTestCase {
+    public ToDatePeriodTests(@Name("TestCase") Supplier<TestCaseSupplier.TestCase> testCaseSupplier) {
+        this.testCase = testCaseSupplier.get();
+    }
+
+    @ParametersFactory
+    public static Iterable<Object[]> parameters() {
+        // TODO multivalue fields
+        String read = "Attribute[channel=0]";
+        List<TestCaseSupplier> suppliers = new ArrayList<>();
+
+        TestCaseSupplier.unary(
+            suppliers,
+            read,
+            validDatePeriod(DataType.DATE_PERIOD),
+            DataType.DATE_PERIOD,
+            bytesRef -> EsqlDataTypeConverter.parseTemporalAmount(((BytesRef) bytesRef).utf8ToString(), DataType.DATE_PERIOD),
+            List.of()
+        );
+
+        for (DataType inputType : AbstractConvertFunction.STRING_TYPES) {
+            TestCaseSupplier.unary(
+                suppliers,
+                read,
+                validDatePeriod(inputType),
+                DataType.DATE_PERIOD,
+                bytesRef -> EsqlDataTypeConverter.parseTemporalAmount(((BytesRef) bytesRef).utf8ToString(), DataType.DATE_PERIOD),
+                List.of()
+            );
+        }
+
+        return parameterSuppliersFromTypedData(
+            errorsForCasesWithoutExamples(anyNullIsNull(true, suppliers), (v, p) -> "date_period or string")
+        );
+    }
+
+    @Override
+    protected Expression build(Source source, List<Expression> args) {
+        return new ToDatePeriod(source, args.get(0));
+    }
+
+    private static List<TestCaseSupplier.TypedDataSupplier> validDatePeriod(DataType dataType) {
+        return List.of(
+            new TestCaseSupplier.TypedDataSupplier("1 day", () -> new BytesRef("1 day"), dataType),
```

Comment:
I think we need to randomize the number here.

I also think we need cover the whole range of allowed literal periods (via randomization). The min and max values for days, weeks, etc. should also best be included in the test cases, as well as 0.

Importantly, whitespaces (leading, trailing and in the middle) should also be randomized.

Same for the durations tests.